In [350]:
# first of all import all the necessary libraries
import pandas as pd
import numpy as np
import plotly.graph_objects as go
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

In [300]:
# now import both training and testing dataset
training_dataset = pd.read_csv('train.csv')
testing_dataset = pd.read_csv('test.csv')

In [301]:
# display training and check
training_dataset.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [302]:
# display testing data
testing_dataset.head(5)

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [303]:
# display columns of training dataset
training_dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [304]:
# shift survived column to the last
survived = training_dataset.pop('Survived')

In [305]:
# now insert that poped column into the last of the dataset
training_dataset.insert(len(training_dataset.columns), 'Survived', survived)
training_dataset

PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Survived  
0           0  
1           1  
2           1  
3           1  
4           0  
..        ...  
886         0  
887         1  
888         0  
889         1  
890         0  

[891 rows x 12 columns]

In [306]:
# know the shape of your dataset
training_dataset.shape

(891, 12)

so there are 891 rows and 12 colums

In [307]:
# now find null values in the dataset
training_dataset.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Survived         0
dtype: int64

so there are 177 missing values in age column and 687 missing values in cabin, and 2 missing values in embarked column

In [308]:
# lets explore cabin column and find unique values in it
training_dataset.Cabin.unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [309]:
# lets solve the issue of cabin by splitting it into two columns namely Has_Cabin and No_Cabin
training_dataset['Has_Cabin'] = training_dataset['Cabin'].notna().astype(int)

In [310]:
training_dataset

PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Survived  Has_Cabin  
0           0          0  
1           1          1  
2           1          0  
3           1          1  
4           0          0  
..        ...        ...  
886         0          0  
887         1          1  
888         0          0  
889         1          1  
890         0          0  

[891 rows x 13 columns]

In [311]:
# so now we have conversted cabin into binary: 1 means has cabin and 0 means no cabin. Lets now remove cabin column
training_dataset.pop('Cabin')

0       NaN
1       C85
2       NaN
3      C123
4       NaN
       ... 
886     NaN
887     B42
888     NaN
889    C148
890     NaN
Name: Cabin, Length: 891, dtype: object

In [312]:
# as of now we have removed cabin colum. now again see which columns contain missing values
training_dataset.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
Survived         0
Has_Cabin        0
dtype: int64

In [313]:
# find out classes in the dataset
training_dataset['Pclass'].unique()

array([3, 1, 2], dtype=int64)

In [314]:
# now lets find which class has most missing values in age column based on gender
print(training_dataset.groupby(['Pclass', 'Sex'])['Age'].apply(lambda x: x.isna().mean()*100))

Pclass  Sex   
1       female     9.574468
        male      17.213115
2       female     2.631579
        male       8.333333
3       female    29.166667
        male      27.089337
Name: Age, dtype: float64


In [315]:
# lets check data types of all the columns
training_dataset.dtypes

PassengerId      int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
Survived         int64
Has_Cabin        int32
dtype: object

In [316]:
# check the name column
titles = training_dataset.Name.str.extract(' ([A-Za-z]+)\. ', expand = False)
titles.unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Don', 'Rev', 'Dr', 'Mme', 'Ms',
       'Major', 'Lady', 'Sir', 'Mlle', 'Col', 'Capt', 'Countess',
       'Jonkheer'], dtype=object)

In [317]:
# i will not drop name column as i wanna use it to create a rank column and then analyze the data based on their ranks
training_dataset['Titles'] = titles

In [318]:
# now drop Name column
training_dataset = training_dataset.drop('Name', axis = 1)

In [319]:
# view dataset
training_dataset.head()

PassengerId  Pclass     Sex   Age  SibSp  Parch            Ticket     Fare  \
0            1       3    male  22.0      1      0         A/5 21171   7.2500   
1            2       1  female  38.0      1      0          PC 17599  71.2833   
2            3       3  female  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  female  35.0      1      0            113803  53.1000   
4            5       3    male  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin Titles  
0        S         0          0     Mr  
1        C         1          1    Mrs  
2        S         1          0   Miss  
3        S         1          1    Mrs  
4        S         0          0     Mr

In [320]:
# now handle sex with one hot encoding
training_dataset = pd.get_dummies(training_dataset, columns = ['Sex'])
training_dataset.head()

PassengerId  Pclass   Age  SibSp  Parch            Ticket     Fare  \
0            1       3  22.0      1      0         A/5 21171   7.2500   
1            2       1  38.0      1      0          PC 17599  71.2833   
2            3       3  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  35.0      1      0            113803  53.1000   
4            5       3  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin Titles  Sex_female  Sex_male  
0        S         0          0     Mr       False      True  
1        C         1          1    Mrs        True     False  
2        S         1          0   Miss        True     False  
3        S         1          1    Mrs        True     False  
4        S         0          0     Mr       False      True

In [321]:
# convert newly created Sex_female and Sex_male into integers
training_dataset.Sex_female = training_dataset['Sex_female'].astype(int)
training_dataset.Sex_male = training_dataset['Sex_male'].astype(int)

In [322]:
# now display training_dataset
training_dataset.head()

PassengerId  Pclass   Age  SibSp  Parch            Ticket     Fare  \
0            1       3  22.0      1      0         A/5 21171   7.2500   
1            2       1  38.0      1      0          PC 17599  71.2833   
2            3       3  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  35.0      1      0            113803  53.1000   
4            5       3  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin Titles  Sex_female  Sex_male  
0        S         0          0     Mr           0         1  
1        C         1          1    Mrs           1         0  
2        S         1          0   Miss           1         0  
3        S         1          1    Mrs           1         0  
4        S         0          0     Mr           0         1

In [323]:
# check dtypes of all columns again
training_dataset.dtypes

PassengerId      int64
Pclass           int64
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
Survived         int64
Has_Cabin        int32
Titles          object
Sex_female       int32
Sex_male         int32
dtype: object

In [324]:
# check the unique values in Ticket, Embarked, and titles
training_dataset.Ticket.nunique()

681

In [325]:
# as there are 681 unique values in ticket column so drop it
training_dataset = training_dataset.drop('Ticket', axis = 1)
training_dataset

PassengerId  Pclass   Age  SibSp  Parch     Fare Embarked  Survived  \
0              1       3  22.0      1      0   7.2500        S         0   
1              2       1  38.0      1      0  71.2833        C         1   
2              3       3  26.0      0      0   7.9250        S         1   
3              4       1  35.0      1      0  53.1000        S         1   
4              5       3  35.0      0      0   8.0500        S         0   
..           ...     ...   ...    ...    ...      ...      ...       ...   
886          887       2  27.0      0      0  13.0000        S         0   
887          888       1  19.0      0      0  30.0000        S         1   
888          889       3   NaN      1      2  23.4500        S         0   
889          890       1  26.0      0      0  30.0000        C         1   
890          891       3  32.0      0      0   7.7500        Q         0   

     Has_Cabin Titles  Sex_female  Sex_male  
0            0     Mr           0         1  
1            1    Mrs           1         0  
2            0   Miss           1         0  
3            1    Mrs           1         0  
4            0     Mr           0         1  
..         ...    ...         ...       ...  
886          0    Rev           0         1  
887          1   Miss           1         0  
888          0   Miss           1         0  
889          1     Mr           0         1  
890          0     Mr           0         1  

[891 rows x 12 columns]

In [326]:
# find number of unique values in embarked and titles
training_dataset[['Embarked', 'Titles']].nunique()

Embarked     3
Titles      17
dtype: int64

In [327]:
# since there are 3 unique values in Embarked column, lets apply one hot encoding
training_dataset = pd.get_dummies(training_dataset, columns = ['Embarked'], dtype = 'int')
training_dataset.head(5)

PassengerId  Pclass   Age  SibSp  Parch     Fare  Survived  Has_Cabin  \
0            1       3  22.0      1      0   7.2500         0          0   
1            2       1  38.0      1      0  71.2833         1          1   
2            3       3  26.0      0      0   7.9250         1          0   
3            4       1  35.0      1      0  53.1000         1          1   
4            5       3  35.0      0      0   8.0500         0          0   

  Titles  Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  
0     Mr           0         1           0           0           1  
1    Mrs           1         0           1           0           0  
2   Miss           1         0           0           0           1  
3    Mrs           1         0           0           0           1  
4     Mr           0         1           0           0           1

In [328]:
# now lets handle Titles as there are 17 unique values
# first check which values are repeated most
training_dataset.Titles.value_counts()

Titles
Mr          517
Miss        182
Mrs         125
Master       40
Dr            7
Rev           6
Mlle          2
Major         2
Col           2
Countess      1
Capt          1
Ms            1
Sir           1
Lady          1
Mme           1
Don           1
Jonkheer      1
Name: count, dtype: int64

In [329]:
# lets again apply one hot encoding as it preserves individual values
training_dataset = pd.get_dummies(training_dataset, columns = ['Titles'], dtype = 'int')
training_dataset.head(5)

PassengerId  Pclass   Age  SibSp  Parch     Fare  Survived  Has_Cabin  \
0            1       3  22.0      1      0   7.2500         0          0   
1            2       1  38.0      1      0  71.2833         1          1   
2            3       3  26.0      0      0   7.9250         1          0   
3            4       1  35.0      1      0  53.1000         1          1   
4            5       3  35.0      0      0   8.0500         0          0   

   Sex_female  Sex_male  ...  Titles_Major  Titles_Master  Titles_Miss  \
0           0         1  ...             0              0            0   
1           1         0  ...             0              0            0   
2           1         0  ...             0              0            1   
3           1         0  ...             0              0            0   
4           0         1  ...             0              0            0   

   Titles_Mlle  Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  \
0            0           0          1           0          0           0   
1            0           0          0           1          0           0   
2            0           0          0           0          0           0   
3            0           0          0           1          0           0   
4            0           0          1           0          0           0   

   Titles_Sir  
0           0  
1           0  
2           0  
3           0  
4           0  

[5 rows x 30 columns]

In [330]:
# now last few things to do like dropping passengerid column
training_dataset = training_dataset.drop(['PassengerId'], axis = 1)

In [331]:
# diplay dataframe
training_dataset.head(3)

Pclass   Age  SibSp  Parch     Fare  Survived  Has_Cabin  Sex_female  \
0       3  22.0      1      0   7.2500         0          0           0   
1       1  38.0      1      0  71.2833         1          1           1   
2       3  26.0      0      0   7.9250         1          0           1   

   Sex_male  Embarked_C  ...  Titles_Major  Titles_Master  Titles_Miss  \
0         1           0  ...             0              0            0   
1         0           1  ...             0              0            0   
2         0           0  ...             0              0            1   

   Titles_Mlle  Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  \
0            0           0          1           0          0           0   
1            0           0          0           1          0           0   
2            0           0          0           0          0           0   

   Titles_Sir  
0           0  
1           0  
2           0  

[3 rows x 29 columns]

In [332]:
# now move Survived column to the last of dataframe
survived = training_dataset['Survived']

training_dataset = training_dataset.drop(['Survived'], axis = 1)

# now insert survived column at the end of dataframe
training_dataset.insert(len(training_dataset.columns), 'Survived', survived)
training_dataset.head(2)

Pclass   Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
0       3  22.0      1      0   7.2500          0           0         1   
1       1  38.0      1      0  71.2833          1           1         0   

   Embarked_C  Embarked_Q  ...  Titles_Master  Titles_Miss  Titles_Mlle  \
0           0           0  ...              0            0            0   
1           1           0  ...              0            0            0   

   Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  Titles_Sir  \
0           0          1           0          0           0           0   
1           0          0           1          0           0           0   

   Survived  
0         0  
1         1  

[2 rows x 29 columns]

In [333]:
# now find dtypes of all data
training_dataset.dtypes

Pclass               int64
Age                float64
SibSp                int64
Parch                int64
Fare               float64
Has_Cabin            int32
Sex_female           int32
Sex_male             int32
Embarked_C           int32
Embarked_Q           int32
Embarked_S           int32
Titles_Capt          int32
Titles_Col           int32
Titles_Countess      int32
Titles_Don           int32
Titles_Dr            int32
Titles_Jonkheer      int32
Titles_Lady          int32
Titles_Major         int32
Titles_Master        int32
Titles_Miss          int32
Titles_Mlle          int32
Titles_Mme           int32
Titles_Mr            int32
Titles_Mrs           int32
Titles_Ms            int32
Titles_Rev           int32
Titles_Sir           int32
Survived             int64
dtype: object

In [334]:
# since no objects
# now find missing values in each columns
training_dataset.isnull().sum()

Pclass               0
Age                177
SibSp                0
Parch                0
Fare                 0
Has_Cabin            0
Sex_female           0
Sex_male             0
Embarked_C           0
Embarked_Q           0
Embarked_S           0
Titles_Capt          0
Titles_Col           0
Titles_Countess      0
Titles_Don           0
Titles_Dr            0
Titles_Jonkheer      0
Titles_Lady          0
Titles_Major         0
Titles_Master        0
Titles_Miss          0
Titles_Mlle          0
Titles_Mme           0
Titles_Mr            0
Titles_Mrs           0
Titles_Ms            0
Titles_Rev           0
Titles_Sir           0
Survived             0
dtype: int64

In [335]:
# in age column we have few missing values, so lets handle them using regiression prediction
# for that first initially find correlation among all variables in training_dataset
corr_matrix = training_dataset.corr(method = 'pearson')

correlation = corr_matrix['Age'].sort_values(ascending = False)
correlation

Age                1.000000
Has_Cabin          0.249732
Titles_Mr          0.206341
Titles_Mrs         0.180278
Titles_Capt        0.103971
Titles_Col         0.103328
Fare               0.096067
Sex_male           0.093254
Titles_Rev         0.085407
Titles_Dr          0.078008
Titles_Major       0.068643
Titles_Sir         0.049794
Titles_Lady        0.047214
Embarked_C         0.036261
Titles_Don         0.026575
Titles_Jonkheer    0.021415
Titles_Countess    0.008516
Titles_Ms         -0.004384
Titles_Mme        -0.014703
Titles_Mlle       -0.020808
Embarked_Q        -0.022405
Embarked_S        -0.032523
Survived          -0.077221
Sex_female        -0.093254
Parch             -0.189119
Titles_Miss       -0.276792
SibSp             -0.308247
Pclass            -0.369226
Titles_Master     -0.398827
Name: Age, dtype: float64

In [336]:
# filter the correlations
correlation[(correlation < -0.2) | (correlation > 0.2)]

Age              1.000000
Has_Cabin        0.249732
Titles_Mr        0.206341
Titles_Miss     -0.276792
SibSp           -0.308247
Pclass          -0.369226
Titles_Master   -0.398827
Name: Age, dtype: float64

In [337]:
# now create a temporary dataset from the original dataset
age_dataset = training_dataset[['Has_Cabin', 'Titles_Mr', 'Titles_Miss', 'SibSp', 'Pclass', 'Titles_Master', 'Age']]

In [340]:
# display age dataset
age_dataset

Has_Cabin  Titles_Mr  Titles_Miss  SibSp  Pclass  Titles_Master   Age
0            0          1            0      1       3              0  22.0
1            1          0            0      1       1              0  38.0
2            0          0            1      0       3              0  26.0
3            1          0            0      1       1              0  35.0
4            0          1            0      0       3              0  35.0
..         ...        ...          ...    ...     ...            ...   ...
886          0          0            0      0       2              0  27.0
887          1          0            1      0       1              0  19.0
888          0          0            1      1       3              0   NaN
889          1          1            0      0       1              0  26.0
890          0          1            0      0       3              0  32.0

[891 rows x 7 columns]

In [343]:
# now split age dataset into training and testing
age_testing = age_dataset[age_dataset['Age'].isnull()]

age_training = age_dataset[~age_dataset['Age'].isnull()]
age_training

Has_Cabin  Titles_Mr  Titles_Miss  SibSp  Pclass  Titles_Master   Age
0            0          1            0      1       3              0  22.0
1            1          0            0      1       1              0  38.0
2            0          0            1      0       3              0  26.0
3            1          0            0      1       1              0  35.0
4            0          1            0      0       3              0  35.0
..         ...        ...          ...    ...     ...            ...   ...
885          0          0            0      0       3              0  39.0
886          0          0            0      0       2              0  27.0
887          1          0            1      0       1              0  19.0
889          1          1            0      0       1              0  26.0
890          0          1            0      0       3              0  32.0

[714 rows x 7 columns]

In [345]:
# now split training into dependent and independent
x_train = age_training[['Has_Cabin', 'Titles_Mr', 'Titles_Miss', 'SibSp', 'Pclass', 'Titles_Master']]
y_train = age_training['Age']

In [347]:
y_train

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
885    39.0
886    27.0
887    19.0
889    26.0
890    32.0
Name: Age, Length: 714, dtype: float64

In [348]:
# split testing into dependent and independent
x_test = age_testing[['Has_Cabin', 'Titles_Mr', 'Titles_Miss', 'SibSp', 'Pclass', 'Titles_Master']]

In [349]:
x_test

Has_Cabin  Titles_Mr  Titles_Miss  SibSp  Pclass  Titles_Master
5            0          1            0      0       3              0
17           0          1            0      0       2              0
19           0          0            0      0       3              0
26           0          1            0      0       3              0
28           0          0            1      0       3              0
..         ...        ...          ...    ...     ...            ...
859          0          1            0      0       3              0
863          0          0            1      8       3              0
868          0          1            0      0       3              0
878          0          1            0      0       3              0
888          0          0            1      1       3              0

[177 rows x 6 columns]

In [351]:
# now use KNN algorithm to predict the remaining age values
lm = LinearRegression()
lm.fit(x_train, y_train)

LinearRegression()

In [352]:
age_missing = lm.predict(x_test)
age_missing

array([29.32955991, 34.97449987, 32.07635213, 29.32955991, 18.3275989 ,
       29.32955991, 41.27702808, 18.3275989 , 29.32955991, 29.32955991,
       29.32955991, 27.25006136, 18.3275989 , 25.1705628 , 40.60973442,
       40.61943982,  4.70784101, 29.32955991, 29.32955991, 18.3275989 ,
       29.32955991, 29.32955991, 29.32955991, 29.32955991, 16.24810035,
       29.32955991, 29.32955991, 16.23839495, 32.07635213, 29.32955991,
       29.32955991, -9.84864888, 43.35652664, 40.61943982,  0.5488439 ,
        1.69161045, 34.97449987, 40.60973442, 29.99685357, 29.32955991,
       18.3275989 , 12.69357146, 27.25006136, 29.32955991, 12.08910323,
       18.3275989 , 16.24810035, 16.24810035, 29.32955991, 43.36623203,
       29.32955991, 18.3275989 , 40.61943982, 18.3275989 , 34.97449987,
       40.60973442, 40.61943982, 40.60973442, 18.3275989 , 25.1705628 ,
       23.96283346, 29.32955991, 29.61747881, 12.69357146, 14.16860179,
       41.28673348, 29.32955991, 29.99685357, 40.60973442, 29.32

In [353]:
# now lets copy original dataset
dataset_copy = training_dataset.copy()

In [354]:
# lets try to put the predicted age values into copied dataset
dataset_copy.loc[dataset_copy['Age'].isnull(), 'Age'] = age_missing

In [355]:
dataset_copy

Pclass      Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
0         3  22.0000      1      0   7.2500          0           0         1   
1         1  38.0000      1      0  71.2833          1           1         0   
2         3  26.0000      0      0   7.9250          0           1         0   
3         1  35.0000      1      0  53.1000          1           1         0   
4         3  35.0000      0      0   8.0500          0           0         1   
..      ...      ...    ...    ...      ...        ...         ...       ...   
886       2  27.0000      0      0  13.0000          0           0         1   
887       1  19.0000      0      0  30.0000          1           1         0   
888       3  16.2481      1      2  23.4500          0           1         0   
889       1  26.0000      0      0  30.0000          1           0         1   
890       3  32.0000      0      0   7.7500          0           0         1   

     Embarked_C  Embarked_Q  ...  Titles_Master  Titles_Miss  Titles_Mlle  \
0             0           0  ...              0            0            0   
1             1           0  ...              0            0            0   
2             0           0  ...              0            1            0   
3             0           0  ...              0            0            0   
4             0           0  ...              0            0            0   
..          ...         ...  ...            ...          ...          ...   
886           0           0  ...              0            0            0   
887           0           0  ...              0            1            0   
888           0           0  ...              0            1            0   
889           1           0  ...              0            0            0   
890           0           1  ...              0            0            0   

     Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  Titles_Sir  \
0             0          1           0          0           0           0   
1             0          0           1          0           0           0   
2             0          0           0          0           0           0   
3             0          0           1          0           0           0   
4             0          1           0          0           0           0   
..          ...        ...         ...        ...         ...         ...   
886           0          0           0          0           1           0   
887           0          0           0          0           0           0   
888           0          0           0          0           0           0   
889           0          1           0          0           0           0   
890           0          1           0          0           0           0   

     Survived  
0           0  
1           1  
2           1  
3           1  
4           0  
..        ...  
886         0  
887         1  
888         0  
889         1  
890         0  

[891 rows x 29 columns]

In [356]:
# null values
dataset_copy.isnull().sum()

Pclass             0
Age                0
SibSp              0
Parch              0
Fare               0
Has_Cabin          0
Sex_female         0
Sex_male           0
Embarked_C         0
Embarked_Q         0
Embarked_S         0
Titles_Capt        0
Titles_Col         0
Titles_Countess    0
Titles_Don         0
Titles_Dr          0
Titles_Jonkheer    0
Titles_Lady        0
Titles_Major       0
Titles_Master      0
Titles_Miss        0
Titles_Mlle        0
Titles_Mme         0
Titles_Mr          0
Titles_Mrs         0
Titles_Ms          0
Titles_Rev         0
Titles_Sir         0
Survived           0
dtype: int64

In [359]:
# now compare
dataset_copy.tail(10)

Pclass      Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
881       3  33.0000      0      0   7.8958          0           0         1   
882       3  22.0000      0      0  10.5167          0           1         0   
883       2  28.0000      0      0  10.5000          0           0         1   
884       3  25.0000      0      0   7.0500          0           0         1   
885       3  39.0000      0      5  29.1250          0           1         0   
886       2  27.0000      0      0  13.0000          0           0         1   
887       1  19.0000      0      0  30.0000          1           1         0   
888       3  16.2481      1      2  23.4500          0           1         0   
889       1  26.0000      0      0  30.0000          1           0         1   
890       3  32.0000      0      0   7.7500          0           0         1   

     Embarked_C  Embarked_Q  ...  Titles_Master  Titles_Miss  Titles_Mlle  \
881           0           0  ...              0            0            0   
882           0           0  ...              0            1            0   
883           0           0  ...              0            0            0   
884           0           0  ...              0            0            0   
885           0           1  ...              0            0            0   
886           0           0  ...              0            0            0   
887           0           0  ...              0            1            0   
888           0           0  ...              0            1            0   
889           1           0  ...              0            0            0   
890           0           1  ...              0            0            0   

     Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  Titles_Sir  \
881           0          1           0          0           0           0   
882           0          0           0          0           0           0   
883           0          1           0          0           0           0   
884           0          1           0          0           0           0   
885           0          0           1          0           0           0   
886           0          0           0          0           1           0   
887           0          0           0          0           0           0   
888           0          0           0          0           0           0   
889           0          1           0          0           0           0   
890           0          1           0          0           0           0   

     Survived  
881         0  
882         0  
883         0  
884         0  
885         0  
886         0  
887         1  
888         0  
889         1  
890         0  

[10 rows x 29 columns]

In [360]:
#
training_dataset.tail(10)

Pclass   Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
881       3  33.0      0      0   7.8958          0           0         1   
882       3  22.0      0      0  10.5167          0           1         0   
883       2  28.0      0      0  10.5000          0           0         1   
884       3  25.0      0      0   7.0500          0           0         1   
885       3  39.0      0      5  29.1250          0           1         0   
886       2  27.0      0      0  13.0000          0           0         1   
887       1  19.0      0      0  30.0000          1           1         0   
888       3   NaN      1      2  23.4500          0           1         0   
889       1  26.0      0      0  30.0000          1           0         1   
890       3  32.0      0      0   7.7500          0           0         1   

     Embarked_C  Embarked_Q  ...  Titles_Master  Titles_Miss  Titles_Mlle  \
881           0           0  ...              0            0            0   
882           0           0  ...              0            1            0   
883           0           0  ...              0            0            0   
884           0           0  ...              0            0            0   
885           0           1  ...              0            0            0   
886           0           0  ...              0            0            0   
887           0           0  ...              0            1            0   
888           0           0  ...              0            1            0   
889           1           0  ...              0            0            0   
890           0           1  ...              0            0            0   

     Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  Titles_Sir  \
881           0          1           0          0           0           0   
882           0          0           0          0           0           0   
883           0          1           0          0           0           0   
884           0          1           0          0           0           0   
885           0          0           1          0           0           0   
886           0          0           0          0           1           0   
887           0          0           0          0           0           0   
888           0          0           0          0           0           0   
889           0          1           0          0           0           0   
890           0          1           0          0           0           0   

     Survived  
881         0  
882         0  
883         0  
884         0  
885         0  
886         0  
887         1  
888         0  
889         1  
890         0  

[10 rows x 29 columns]

In [361]:
# seems it is correct. now copy the dataset_copy as final dataset and then apply final model for survival rate
final_dataset = dataset_copy.copy()

In [367]:
final_dataset.head()

Pclass   Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
0       3  22.0      1      0   7.2500          0           0         1   
1       1  38.0      1      0  71.2833          1           1         0   
2       3  26.0      0      0   7.9250          0           1         0   
3       1  35.0      1      0  53.1000          1           1         0   
4       3  35.0      0      0   8.0500          0           0         1   

   Embarked_C  Embarked_Q  ...  Titles_Master  Titles_Miss  Titles_Mlle  \
0           0           0  ...              0            0            0   
1           1           0  ...              0            0            0   
2           0           0  ...              0            1            0   
3           0           0  ...              0            0            0   
4           0           0  ...              0            0            0   

   Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  Titles_Sir  \
0           0          1           0          0           0           0   
1           0          0           1          0           0           0   
2           0          0           0          0           0           0   
3           0          0           1          0           0           0   
4           0          1           0          0           0           0   

   Survived  
0         0  
1         1  
2         1  
3         1  
4         0  

[5 rows x 29 columns]

In [369]:
# split the final dataset into dependent and independent sets
x = final_dataset.iloc[:, :-1]
y = final_dataset.iloc[:, -1]
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [379]:
# now split x and y into trainig and testing

# make a splitn index
split_index = int(len(x)*0.85)

# split x according to the split index
x_train = x.iloc[:split_index , :]
y_train = x.iloc[split_index:, :]

# now split y while considering split index
x_test = y.iloc[:split_index]
y_test = y.iloc[split_index:]

In [380]:
# print lengths of all training and testing datasets
print(len(x_train))

Pclass    Age  SibSp  Parch     Fare  Has_Cabin  Sex_female  Sex_male  \
0         3  22.00      1      0   7.2500          0           0         1   
1         1  38.00      1      0  71.2833          1           1         0   
2         3  26.00      0      0   7.9250          0           1         0   
3         1  35.00      1      0  53.1000          1           1         0   
4         3  35.00      0      0   8.0500          0           0         1   
..      ...    ...    ...    ...      ...        ...         ...       ...   
752       3  33.00      0      0   9.5000          0           0         1   
753       3  23.00      0      0   7.8958          0           0         1   
754       2  48.00      1      2  65.0000          0           1         0   
755       2   0.67      1      1  14.5000          0           0         1   
756       3  28.00      0      0   7.7958          0           0         1   

     Embarked_C  Embarked_Q  ...  Titles_Major  Titles_Master  Titles_Miss  \
0             0           0  ...             0              0            0   
1             1           0  ...             0              0            0   
2             0           0  ...             0              0            1   
3             0           0  ...             0              0            0   
4             0           0  ...             0              0            0   
..          ...         ...  ...           ...            ...          ...   
752           0           0  ...             0              0            0   
753           0           0  ...             0              0            0   
754           0           0  ...             0              0            0   
755           0           0  ...             0              1            0   
756           0           0  ...             0              0            0   

     Titles_Mlle  Titles_Mme  Titles_Mr  Titles_Mrs  Titles_Ms  Titles_Rev  \
0              0           0          1           0          0           0   
1              0           0          0           1          0           0   
2              0           0          0           0          0           0   
3              0           0          0           1          0           0   
4              0           0          1           0          0           0   
..           ...         ...        ...         ...        ...         ...   
752            0           0          1           0          0           0   
753            0           0          1           0          0           0   
754            0           0          0           1          0           0   
755            0           0          0           0          0           0   
756            0           0          1           0          0           0   

     Titles_Sir  
0             0  
1             0  
2             0  
3             0  
4             0  
..          ...  
752           0  
753           0  
754           0  
755           0  
756           0  

[757 rows x 28 columns]